trace_replay/trace_record_handler.cc

```diff
@@ -28,33 +30,72 @@ TraceExecutionHandler::TraceExecutionHandler(
 
 TraceExecutionHandler::~TraceExecutionHandler() { cf_map_.clear(); }
 
-Status TraceExecutionHandler::Handle(const WriteQueryTraceRecord& record) {
+Status TraceExecutionHandler::Handle(
+    const WriteQueryTraceRecord& record,
+    std::unique_ptr<TraceRecordResult>* result) {
+  if (result != nullptr) {
+    result->reset(nullptr);
+  }
+  uint64_t start = clock_->NowMicros();
+
   WriteBatch batch(record.GetWriteBatchRep().ToString());
-  return db_->Write(write_opts_, &batch);
+  Status s = db_->Write(write_opts_, &batch);
+
+  uint64_t end = clock_->NowMicros();
+
+  if (s.ok() && result != nullptr) {
+    result->reset(new StatusOnlyTraceExecutionResult(s, start, end,
+                                                     record.GetTraceType()));
+  }
+
+  return s;
 }
 
-Status TraceExecutionHandler::Handle(const GetQueryTraceRecord& record) {
+Status TraceExecutionHandler::Handle(
+    const GetQueryTraceRecord& record,
+    std::unique_ptr<TraceRecordResult>* result) {
+  if (result != nullptr) {
+    result->reset(nullptr);
+  }
   auto it = cf_map_.find(record.GetColumnFamilyID());
   if (it == cf_map_.end()) {
     return Status::Corruption("Invalid Column Family ID.");
   }
 
+  uint64_t start = clock_->NowMicros();
+
   std::string value;
   Status s = db_->Get(read_opts_, it->second, record.GetKey(), &value);
 
-  // Treat not found as ok and return other errors.
-  return s.IsNotFound() ? Status::OK() : s;
+  uint64_t end = clock_->NowMicros();
+
+  // Treat not found as ok, return other errors.
+  if (!s.ok() && !s.IsNotFound()) {
+    return s;
+  }
+
+  if (result != nullptr) {
+    // Report the actual opetation status in TraceExecutionResult
+    result->reset(new SingleValueTraceExecutionResult(
+        s, std::move(value), start, end, record.GetTraceType()));
+  }
+  return Status::OK();
 }
 
 Status TraceExecutionHandler::Handle(
-    const IteratorSeekQueryTraceRecord& record) {
+    const IteratorSeekQueryTraceRecord& record,
+    std::unique_ptr<TraceRecordResult>* result) {
+  if (result != nullptr) {
+    result->reset(nullptr);
+  }
   auto it = cf_map_.find(record.GetColumnFamilyID());
   if (it == cf_map_.end()) {
     return Status::Corruption("Invalid Column Family ID.");
   }
 
-  Iterator* single_iter = db_->NewIterator(read_opts_, it->second);
+  uint64_t start = clock_->NowMicros();
 
+  Iterator* single_iter = db_->NewIterator(read_opts_, it->second);
```

Comment:
I think one of the benefits of moving timing down to the TraceRecord handler is we can exclude setup/teardown like iterator creation/destruction. Should we move start and end accordingly so it only measures `Seek()` / `SeekForPrev()` since that's what the optype indicates? (Maybe in the future we have separate iterator creation/destruction record types and then this operation would just reference an already existing iterator.)

Reply:
Done